test/Core.Test/Auth/UserFeatures/TdeOffboarding/TdeOffboardingPasswordCommandTests.cs

```diff
@@ -0,0 +1,98 @@
+            .Returns(IdentityResult.Success);
+
+        orgUserDetails.UseSso = true;
+        sutProvider.GetDependency<IOrganizationUserRepository>()
+            .GetManyDetailsByUserAsync(user.Id)
+            .Returns(new List<OrganizationUserOrganizationDetails> { orgUserDetails });
+
+        sutProvider.GetDependency<ISsoUserRepository>()
+            .GetByUserIdOrganizationIdAsync(orgUserDetails.OrganizationId, user.Id)
+            .Returns(ssoUser);
+
+        var ssoConfig = new SsoConfig();
+        var ssoConfigData = ssoConfig.GetData();
+        ssoConfigData.MemberDecryptionType = MemberDecryptionType.MasterPassword;
+        ssoConfig.SetData(ssoConfigData);
+        sutProvider.GetDependency<ISsoConfigRepository>()
+            .GetByOrganizationIdAsync(orgUserDetails.OrganizationId)
+            .Returns(ssoConfig);
+
+        // Act
+        var result = await sutProvider.Sut.UpdateTdeOffboardingPasswordAsync(user, masterPassword, key, hint);
+
+        // Assert
+        Assert.Equal(IdentityResult.Success, result);
+    }
+
+    [Theory]
+    [BitAutoData]
+    public async Task TdeOffboardingPasswordCommand_RejectWithTdeEnabled(SutProvider<TdeOffboardingPasswordCommand> sutProvider,
+        User user, string masterPassword, string key, string hint, OrganizationUserOrganizationDetails orgUserDetails, SsoUser ssoUser)
+    {
+        // Arrange
+        user.MasterPassword = null;
+
+        sutProvider.GetDependency<IUserService>()
+            .UpdatePasswordHash(Arg.Any<User>(), Arg.Any<string>(), true, false)
+            .Returns(IdentityResult.Success);
+
+        orgUserDetails.UseSso = true;
+        sutProvider.GetDependency<IOrganizationUserRepository>()
+            .GetManyDetailsByUserAsync(user.Id)
+            .Returns(new List<OrganizationUserOrganizationDetails> { orgUserDetails });
+
+        sutProvider.GetDependency<ISsoUserRepository>()
+            .GetByUserIdOrganizationIdAsync(orgUserDetails.OrganizationId, user.Id)
+            .Returns(ssoUser);
+
+        var ssoConfig = new SsoConfig();
+        var ssoConfigData = ssoConfig.GetData();
+        ssoConfigData.MemberDecryptionType = MemberDecryptionType.TrustedDeviceEncryption;
+        ssoConfig.SetData(ssoConfigData);
+        sutProvider.GetDependency<ISsoConfigRepository>()
+            .GetByOrganizationIdAsync(orgUserDetails.OrganizationId)
+            .Returns(ssoConfig);
+
+        await Assert.ThrowsAsync<BadRequestException>(() => sutProvider.Sut.UpdateTdeOffboardingPasswordAsync(user, masterPassword, key, hint));
+    }
+
+
+    [Theory]
+    [BitAutoData]
+    public async Task TdeOffboardingPasswordCommand_RejectWithMasterPassword(SutProvider<TdeOffboardingPasswordCommand> sutProvider,
+        User user, string masterPassword, string key, string hint)
+    {
+        await Assert.ThrowsAsync<BadRequestException>(() => sutProvider.Sut.UpdateTdeOffboardingPasswordAsync(user, masterPassword, key, hint));
+    }
```

Comment:
question: I'm not sure I understand what this is testing, or how its throwing. Can you add some clarity on this one?

Reply:
Added a comment explaining this in e58b4c4. The idea is that TDE offboarding should only be possible when the user is actually a passwordless TDE user, that is, has no master password set. If there is a MP set, then the request should be rejected.